datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -181,24 +185,30 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
             expr = expr.rewrite(&mut Canonicalizer::new()).data()?
         }
 
-        // TODO iterate until no changes are made during rewrite
-        // (evaluating constants can enable new simplifications and
-        // simplifications can enable new constant evaluation)
-        // https://github.com/apache/datafusion/issues/1160
-        expr.rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
-            .rewrite(&mut guarantee_rewriter)
-            .data()?
-            // run both passes twice to try an minimize simplifications that we missed
-            .rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
+        let mut i = 0;
+        loop {
```

Comment:
I think you can write this loop more concisely using the `Transformed::transform_data` like this (BTW this is confusing, I made a PR jsut today to try and clarify it here https://github.com/apache/datafusion/pull/10355)

```rust
        let mut i = 0;
        loop {
            let result = expr.rewrite(&mut const_evaluator)?
                .transform_data(|expr| expr.rewrite(&mut simplifier))?
                .transform_data(|expr|  expr.rewrite(&mut guarantee_rewriter))?
                // shorten inlist should be started after other inlist rules are applied
                .transform_data(|expr| expr.rewrite(&mut shorten_in_list_simplifier))?;

            i += 1;
            if !result.transformed || i >= self.max_simplifier_iterations {
                return Ok(result.data);
            }
        }
    }
```